File: Sources/SSHTunnelManager/Models/KnownHostsStore.swift
import Foundation
import Combine

/// One entry (line) in `~/.ssh/known_hosts`.
struct KnownHostEntry: Identifiable, Hashable {
    let id = UUID()
    /// The exact original line, used to rewrite the file when removing.
    let rawLine: String
    /// A friendly host label (or "Hashed host" for a `|1|…` hashed entry).
    let hostLabel: String
    /// The key type (e.g. `ssh-ed25519`, `ecdsa-sha2-nistp256`), when parseable.
    let keyType: String
    /// Whether the host field is hashed (so it can't be shown in the clear).
    let isHashed: Bool
}

/// Reads and edits `~/.ssh/known_hosts` so a changed/stale host key can be
/// removed from inside the app instead of dropping to a shell.
final class KnownHostsStore: ObservableObject {
    @Published private(set) var entries: [KnownHostEntry] = []
    @Published var errorMessage: String?

    static var path: String {
        (NSHomeDirectory() as NSString).appendingPathComponent(".ssh/known_hosts")
    }

    var fileExists: Bool { FileManager.default.fileExists(atPath: KnownHostsStore.path) }

    func reload() {
        errorMessage = nil
        guard fileExists else { entries = []; return }
        guard let text = try? String(contentsOfFile: KnownHostsStore.path, encoding: .utf8) else {
            entries = []
            errorMessage = "Couldn't read ~/.ssh/known_hosts."
            return
        }
        entries = KnownHostsStore.parse(text)
    }

    /// Parse known_hosts text into per-line entries (blank/comment lines skipped).
    static func parse(_ text: String) -> [KnownHostEntry] {
        text.split(whereSeparator: { $0 == "\n" || $0 == "\r" }).compactMap { rawSlice in
            let raw = String(rawSlice)
            let trimmed = raw.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("#") else { return nil }
            let fields = trimmed.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
            guard fields.count >= 2 else { return nil }
            // A leading @marker (e.g. @cert-authority, @revoked) shifts the fields.
            var hostField = fields[0]
            var keyType = fields[1]
            if hostField.hasPrefix("@"), fields.count >= 3 {
                hostField = fields[1]
                keyType = fields[2]
            }
            let isHashed = hostField.hasPrefix("|")
            let hostLabel = isHashed
                ? "Hashed host"
                : hostField.split(separator: ",").map(prettyHost).joined(separator: ", ")
            return KnownHostEntry(rawLine: raw, hostLabel: hostLabel,
                                  keyType: keyType, isHashed: isHashed)
        }
    }

    /// Strip the `[host]:port` brackets ssh uses for non-standard ports.
    private static func prettyHost(_ token: Substring) -> String {
        let s = String(token)
        if s.hasPrefix("["), let close = s.firstIndex(of: "]") {
            let host = s[s.index(after: s.startIndex)..<close]
            let after = s[s.index(after: close)...]
            let port = after.hasPrefix(":") ? after.dropFirst() : ""
            return port.isEmpty ? String(host) : "\(host):\(port)"
        }
        return s
    }

    /// Remove one entry by rewriting the file without its exact line (works for
    /// hashed entries too, which `ssh-keygen -R` can't target by name).
    func remove(_ entry: KnownHostEntry) {
        rewrite(removing: [entry])
    }

    func remove(_ toRemove: [KnownHostEntry]) {
        rewrite(removing: toRemove)
    }

    private func rewrite(removing toRemove: [KnownHostEntry]) {
        guard fileExists,
              let text = try? String(contentsOfFile: KnownHostsStore.path, encoding: .utf8) else { return }
        let doomed = Set(toRemove.map(\.rawLine))
        // Preserve every other line (including blanks / comments) verbatim.
        let kept = text.split(separator: "\n", omittingEmptySubsequences: false)
            .filter { !doomed.contains(String($0)) }
        let rebuilt = kept.joined(separator: "\n")
        do {
            try rebuilt.write(toFile: KnownHostsStore.path, atomically: true, encoding: .utf8)
            reload()
        } catch {
            errorMessage = "Couldn't update known_hosts: \(error.localizedDescription)"
        }
    }
}

File: cross-platform/src/RemoteStuff/Services/KnownHostsStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace RemoteStuff.Services;

/// <summary>One entry (line) in <c>~/.ssh/known_hosts</c>.</summary>
public sealed class KnownHostEntry
{
    public required string RawLine { get; init; }
    public required string HostLabel { get; init; }
    public required string KeyType { get; init; }
    public bool IsHashed { get; init; }
}

/// <summary>
/// Reads and edits <c>~/.ssh/known_hosts</c> so a changed/stale host key can be
/// removed from inside the app. A cross-platform port of the macOS store.
/// </summary>
public sealed class KnownHostsStore
{
    public static string Path =>
        System.IO.Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".ssh", "known_hosts");

    public bool FileExists => File.Exists(Path);
    public string? ErrorMessage { get; private set; }
    public List<KnownHostEntry> Entries { get; private set; } = new();

    public void Reload()
    {
        ErrorMessage = null;
        if (!FileExists) { Entries = new(); return; }
        try
        {
            Entries = Parse(File.ReadAllText(Path));
        }
        catch
        {
            Entries = new();
            ErrorMessage = "Couldn't read ~/.ssh/known_hosts.";
        }
    }

    /// <summary>Parse known_hosts text into per-line entries (blank/comment lines skipped).</summary>
    public static List<KnownHostEntry> Parse(string text)
    {
        var result = new List<KnownHostEntry>();
        foreach (var raw in text.Split(new[] { '\n', '\r' }))
        {
            var trimmed = raw.Trim();
            if (trimmed.Length == 0 || trimmed.StartsWith("#")) continue;
            var fields = trimmed.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
            if (fields.Length < 2) continue;

            var hostField = fields[0];
            var keyType = fields[1];
            if (hostField.StartsWith("@") && fields.Length >= 3)
            {
                hostField = fields[1];
                keyType = fields[2];
            }

            var isHashed = hostField.StartsWith("|");
            var hostLabel = isHashed
                ? "Hashed host"
                : string.Join(", ", hostField.Split(',').Select(PrettyHost));
            result.Add(new KnownHostEntry
            {
                RawLine = raw,
                HostLabel = hostLabel,
                KeyType = keyType,
                IsHashed = isHashed
            });
        }
        return result;
    }

    /// <summary>Strip the <c>[host]:port</c> brackets ssh uses for non-standard ports.</summary>
    private static string PrettyHost(string token)
    {
        if (token.StartsWith("[") && token.Contains(']'))
        {
            var close = token.IndexOf(']');
            var host = token[1..close];
            var after = token[(close + 1)..];
            var port = after.StartsWith(":") ? after[1..] : "";
            return port.Length == 0 ? host : $"{host}:{port}";
        }
        return token;
    }

    public void Remove(KnownHostEntry entry) => Rewrite(new[] { entry });
    public void Remove(IEnumerable<KnownHostEntry> entries) => Rewrite(entries);

    private void Rewrite(IEnumerable<KnownHostEntry> toRemove)
    {
        if (!FileExists) return;
        try
        {
            var text = File.ReadAllText(Path);
            var doomed = toRemove.Select(e => e.RawLine).ToHashSet();
            var kept = text.Split('\n').Where(line => !doomed.Contains(line.TrimEnd('\r')) && !doomed.Contains(line));
            File.WriteAllText(Path, string.Join("\n", kept));
            Reload();
        }
        catch (Exception ex)
        {
            ErrorMessage = "Couldn't update known_hosts: " + ex.Message;
        }
    }
}
